Skip test which only works in_tree

milljm opened the pull request:
Skip the following test, if being tested out-of-tree. Added reference to the issue where this test should have been included with the rest.

This should fix the Conda MOOSE (ARM Mac) target on next.

Closes #25293

Commit:
Skip test which only works in_tree

Skip the following test, if being tested out-of-tree.
Added reference to the issue where this test should have been included
with the rest.

Closes #25293

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1722654/) on 6e46696 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25294/site)

This comment will be updated on new commits.

Commit:
Additional tests which work in_tree only

in_tree 'only' is a bit harsh. These tests seem to pass out-of-tree
except on Apple Silicon machines. Making it a sorta strange case.

moosebuild commented:
All [jobs](https://civet.inl.gov/event/140837/) on 7abf6b8 : canceled by @milljm

Commit:
Another one

This test didn't fail among the last log. Not sure how all tests
are not being run

Oh! Ugh. `set +e` is not being set in run_tests (conda/moose/run_tests.sh)

lindsayad commented on modules/contact/test/tests/mortar_restart/tests:
```diff
@@ -25,5 +25,7 @@
     prereq = 'frictional_bouncing_block_action_restart_1'
     valgrind = 'heavy'
     method = '!dbg'
+    # skip test if test is being run out-of-tree. Issue Ref: #25279
+    installation_type = in_tree
```
It would be nice to understand what are criteria for tests being run out of tree. I don't know why this one won't run out of tree. Is it because there's a prereq? Because it requires a file mesh?

milljm replied:
I'd be happy to add real reasons rather than blanket ones... but I don't know much about what generates these errors. Here is the produced error for `2d-finite-strain-steady/tests`:

```pre
2d-finite-strain-steady.test: *** ERROR ***
2d-finite-strain-steady.test: ADFParser::JITCompile() failed. Evaluation not possible.
2d-finite-strain-steady.test: 
2d-finite-strain-steady.test: Stack frames: 19
2d-finite-strain-steady.test: 0: 0   libmesh_opt.0.dylib                 0x0000000108827ce0 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char>>&) + 1044
2d-finite-strain-steady.test: 1: 1   libmoose-opt.0.dylib                0x00000001050aff58 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) + 680
2d-finite-strain-steady.test: 2: 2   libmoose-opt.0.dylib                0x000000010437df44 void mooseError<char const (&) [57]>(char const (&) [57]) + 216
2d-finite-strain-steady.test: 3: 3   libmoose-opt.0.dylib                0x0000000104ec2394 ADFParser::JITCompile() + 948
2d-finite-strain-steady.test: 4: 4   libmoose-opt.0.dylib                0x000000010441899c ParsedMaterialHelper<true>::functionsOptimize() + 48
2d-finite-strain-steady.test: 5: 5   libmoose-opt.0.dylib                0x000000010441af48 ParsedMaterialHelper<true>::functionsPostParse() + 28
2d-finite-strain-steady.test: 6: 6   libmoose-opt.0.dylib                0x000000010441adc8 ParsedMaterialHelper<true>::functionParse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<PostprocessorName, std::__1::allocator<PostprocessorName>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<double, std::__1::allocator<double>> const&) + 1988
2d-finite-strain-steady.test: 7: 7   libmoose-opt.0.dylib                0x0000000104463844 ParsedMaterialTempl<true>::ParsedMaterialTempl(InputParameters const&) + 480
2d-finite-strain-steady.test: 8: 8   libmoose-opt.0.dylib                0x00000001044635e8 RegistryEntry<ParsedMaterialTempl<true>>::build(InputParameters const&) + 64
2d-finite-strain-steady.test: 9: 9   libmoose-opt.0.dylib                0x0000000105031b84 Factory::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters const&, unsigned int, bool) + 344
2d-finite-strain-steady.test: 10: 10  libmoose-opt.0.dylib                0x0000000104815990 std::__1::shared_ptr<MaterialBase> Factory::create<MaterialBase>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters const&, unsigned int) + 56
2d-finite-strain-steady.test: 11: 11  libmoose-opt.0.dylib                0x0000000104814268 FEProblemBase::addMaterialHelper(std::__1::vector<MaterialWarehouse*, std::__1::allocator<MaterialWarehouse*>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters&) + 564
2d-finite-strain-steady.test: 12: 12  libmoose-opt.0.dylib                0x0000000104813fd8 FEProblemBase::addMaterial(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters&) + 112
2d-finite-strain-steady.test: 13: 13  libmoose-opt.0.dylib                0x0000000104b37de0 ActionWarehouse::executeActionsWithAction(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) + 1160
2d-finite-strain-steady.test: 14: 14  libmoose-opt.0.dylib                0x0000000104b659bc ActionWarehouse::executeAllActions() + 244
2d-finite-strain-steady.test: 15: 15  libmoose-opt.0.dylib                0x0000000105087e58 MooseApp::runInputFile() + 152
2d-finite-strain-steady.test: 16: 16  libmoose-opt.0.dylib                0x000000010508468c MooseApp::run() + 1364
2d-finite-strain-steady.test: 17: 17  combined-opt                        0x0000000100737b4c main + 164
2d-finite-strain-steady.test: 18: 18  dyld                                0x000000019f16ff28 start + 2236
2d-finite-strain-steady.test: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
2d-finite-strain-steady.test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
2d-finite-strain-steady.test: :
2d-finite-strain-steady.test: system msg for write_line failure : Bad file descriptor
2d-finite-strain-steady.test: 
2d-finite-strain-steady.test: 
2d-finite-strain-steady.test: Exit Code: 1
2d-finite-strain-steady.test: ################################################################################
2d-finite-strain-steady.test: Tester failed, reason: ERRMSG
```

cticenhour replied:
Pinging @dschwen for his perspective on JIT in these cases, or what we might do to allow these tests on all installation types.

loganharbour replied:
https://github.com/idaholab/moose/blob/a52b2572d064d7bf7c58825bf243f3f59104eea1/framework/src/utils/ADFParser.C#L50

FParser isn't available in the installed binary it seems?

milljm replied:
well this is odd:

```pre
    SKIP umat.gap_heat_transfer_umat [test requires "IN_TREE" binary]
<trimmed>
      OK feature_volume_fraction.volume_fraction_test
      OK adaptive_timestepping.test_function_change
      OK combined_plasticity_temperature.temp_dep_yield
      OK restart-transient-from-ss-with-stateful.ss
invOpt_nonlinear.BFGS: Working Directory: /private/var/folders/1j/j662yjdx2zb1gspprj00hd70664rrb/T/tmp.eCRXSczM/combined/combined/invOpt_nonlinear
invOpt_nonlinear.BFGS: Running command: /Users/milljm/moose_installed/share/moose/python/TestHarness/../../../../bin/combined-opt -i main.i Executioner/tao_solver=taolmvm --error --error-override --no-gdb-backtrace
invOpt_nonlinear.BFGS: 
```
I am not sure why `invOpt_nonlinear` is not being skipped like it is supposed to.

milljm replied:
I don't suppose that JIT compiles (and fails) before evaluating whether or not the test should run or not?

milljm replied:
damn, the local build on my machine still fails.

milljm replied:
Well I am at a loss.

```pre
moose                     2023.08.19              build_0    https://conda.software.inl.gov/public
```
...version works. I'll need to dig a little bit to understand what might be different. I am sure this is going to be finicky conda-build environment.

lindsayad replied:
I've added https://github.com/idaholab/moose/issues/25293#issuecomment-1692359189. Perhaps we can add reasons as we determine them there

dschwen replied:
I'm a bit late here, but the whole `ADRealMonolithic.h` stuff was created to make JIT compiling work for binary builds. This institutional short term memory loss is a bit concerning. How can we avoid this?

dschwen replied:
FYI, this is documented under modules/phase_field/doc/content/modules/phase_field/FunctionMaterials/JITCompilation.md

loganharbour replied:
Resolving to merge a fix for now, but we should continue this discussion

dschwen replied:
> This is a little severe I think.

Ah, I see where the check-in came from :-D. This is not meant as a personal criticism of anyone involved. I'm just observing a structural problem that needs to be addressed IMO. The documentation was quite apparently either not enough, or not discoverable enough. Design docs are supposed to prevent these types of things from happening. 

I suppose we could link explicitly to the relevant design doc from an error message (`ADFParser::JITCompile() failed. Evaluation not possible. Check JITCompilation.md to see if the environment is set up correctly.`).

milljm replied:
This could be an environment 'thing', but seeing as we've been failing on Civet, where we blow away the environment every time... I gotta ask:

Using `moose` that is available in public _now_, everything works:
```bash
❯ mamba create -n prebuilt moose
❯ mamba activate prebuilt
❯ moose --copy-inputs optimization
❯ cd moose/optimization
❯ moose --run -j 12 --re=nonlinear
Working Directory: /private/tmp/testing/moose/optimization
Running Command: /Users/milljm/mambaforge3/envs/testing/moose/bin/moose_test_runner -j 12 --re nonlinear
      OK executioners/steady_and_adjoint.steady_and_adjoint/nonlinear
      OK executioners/transient_and_adjoint.transient_and_adjoint/nonlinear
      OK optimizationreporter/constant_heat_source.constant_heat_source/nonlinear
      OK optimizationreporter/nonlinear_material.nonlinear_material
--------------------------------------------------------------------------------------------------------------
Ran 4 tests in 1.7 seconds. Average test time 1.0 seconds, maximum test time 1.6 seconds.
```

using `moose` that I am building on HEAD, is failing for the same tests. With or without setting `MOOSE_ADFPARSER_JIT_INCLUDE`.

I'd also like to point out that in public MOOSE as demonstrated above, we are NOT setting `MOOSE_ADFPARSER_JIT_INCLUDE`!

These errors are also only happening on Apple Silicon machines.

Full log of a single test failure will be attached...
[log.txt](https://github.com/idaholab/moose/files/12457657/log.txt)

milljm replied:
Not sure if it matters, but here is a log of the same test, after setting MOOSE_ADFPARSER_JIT_INCLUDE (still fails)

```pre
Final Test Results:
--------------------------------------------------------------------------------------------------------------
      OK optimizationreporter/constant_heat_source.constant_heat_source/nonlinear
    FAIL optimizationreporter/nonlinear_material.nonlinear_material FAILED (ERRMSG)
    FAIL executioners/transient_and_adjoint.transient_and_adjoint/nonlinear FAILED (ERRMSG)
    FAIL executioners/steady_and_adjoint.steady_and_adjoint/nonlinear FAILED (ERRMSG)
--------------------------------------------------------------------------------------------------------------
Ran 4 tests in 1.4 seconds. Average test time 1.2 seconds, maximum test time 1.3 seconds.
1 passed, 0 skipped, 0 pending, 3 FAILED


*** ERROR ***
Run failed

Stack frames: 7
0: 0   libmesh_opt.0.dylib                 0x0000000108607ce0 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char>>&) + 1044
1: 1   libmoose-opt.0.dylib                0x0000000104e8f940 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) + 680
2: 2   libmoose-opt.0.dylib                0x0000000104e5e7f8 void mooseError<char const (&) [11]>(char const (&) [11]) + 216
3: 3   libmoose-opt.0.dylib                0x0000000104e5e410 MooseApp::runInputs() const + 1044
4: 4   libmoose-opt.0.dylib                0x0000000104e63ee0 MooseApp::run() + 960
5: 5   combined-opt                        0x0000000100517b4c main + 164
6: 6   dyld                                0x000000019ef7ff28 start + 2236
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
❯ echo $MOOSE_ADFPARSER_JIT_INCLUDE
/Users/milljm/mambaforge3/envs/prebuilt/moose/include/moose/ADRealMonolithic.h
❯ file $MOOSE_ADFPARSER_JIT_INCLUDE
/Users/milljm/mambaforge3/envs/prebuilt/moose/include/moose/ADRealMonolithic.h: c program text, ASCII text
```
[log.txt](https://github.com/idaholab/moose/files/12457682/log.txt)

dschwen replied:
This error is super weird IMO. The two declarations are in the _same_ file. I wonder if there is a missing preprocessor symbol that switches between them.

lindsayad replied:
> I'm just observing a structural problem that needs to be addressed IMO. The documentation was quite apparently either not enough, or not discoverable enough. Design docs are supposed to prevent these types of things from happening.
>
> I suppose we could link explicitly to the relevant design doc from an error message (ADFParser::JITCompile() failed. Evaluation not possible. Check JITCompilation.md to see if the environment is set up correctly.).

Just commenting quickly on this. Yea there are some flaws. Adding the relevant markdown file in the error would definitely make it more likely that we end up looking at it. I did test searching for JIT compilation on the website and the markdown file came up at the top with a couple different searches, so we probably should have come across it this time even without the explicit reference in the error. But sometimes unless I'm explicitly told something I may be too lazy to do the work 😦

Commit:
Add JIT header support

Oddly, I thought this was already being set. But probably, I was
getting our HPC binaries confused with Conda's.

Commit:
Use PREFIX, not CONDA_PREFIX...

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1723814/) on 0f75432 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a52b25</th>
      <th colspan=4>#25294 0f7543</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.65%</td>
      <td>85.66%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90273</td>
      <td>90277</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15119</td>
      <td>15115</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25294/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25294/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25294/coverage/xfem)

</details>



This comment will be updated on new commits.

Commit:
All failing due to JIT

Even with proper environment variables, JIT seems to be failing
on Apple Silicon.

milljm commented:
@loganharbour  This should fix next. Unless something new crops up.
